Contrast support, MCU Controlled PPS out and other stuff

fredzo opened the pull request:
Hi Daniel,

I took the liberty of making some changes on your excellent work on this alternative firmware for the GPSDO.
It seemed that you where no longer active on eevblog forum so I tried to feel some users expectations (see here: https://www.eevblog.com/forum/testgear/10mhz-gpsdo-by-bh3sap/msg5815903/#msg5815903 and preceding discussions).
Here are some of the changes made:
- Documentation update
- Contrast setting
- Comprehensive menu system with:
    - PPB sub menu to see mean and instant PPB values along with measured MCU clock frequency and PWM correction
    - GPS sub menu to display various GPS ingormation (latitude, longitude, altitude, HDOP etc.)
    - PPS sub menu to monitor and configure MCU controlled PPS output
    - Version menu to show firmware revision
- LED1 blink every second (as on the original firmware)
- Fixed PWM save menu
- 'X' icon blinks when the GPS module is not locked

Please let me know if you have any comment.

Best,

Frederic.

fredzo committed:
Added contrast control.

fredzo committed:
No sattelite indicator is now blikning every second.

fredzo committed:
Added gga frame count screen.

fredzo committed:
Added MCU PPS OUT on pin PB1

fredzo committed:
Fixed PPS output frequency.

fredzo committed:
Added GPS menu

fredzo committed:
Added 1 PPS  LED1 blnk.

fredzo committed:
Added ppb sub-menu.

fredzo committed:
Added datasheets.

fredzo committed:
Added gitignore

fredzo committed:
Added datasheet.

fredzo committed:
Fixed menu navigation.

fredzo committed:
Fixed encoder handling.

fredzo committed:
Update README.md

fredzo committed:
Update README.md

fredzo committed:
Added firmware version screen.

fredzo committed:
Merge remote-tracking branch 'origin/main'

fredzo committed:
Added docs

fredzo committed:
Update README.md

fredzo committed:
Fixed file name

fredzo committed:
Update README.md

fredzo committed:
Update README.md

fredzo committed:
Update README.md

fredzo committed:
Update README.md

fredzo committed:
Added doc

fredzo committed:
Update README.md

fredzo committed:
Added doc

fredzo committed:
Update README.md

fredzo committed:
Update README.md

fredzo committed:
Added doc

fredzo committed:
Update README.md

fredzo committed:
doc

fredzo committed:
Merge remote-tracking branch 'origin/main'

fredzo committed:
Update README.md

fredzo committed:
Added PPS menu

fredzo committed:
Update README.md

fredzo committed:
Update README.md

dankar commented:
Awesome! I actually thought I was subscribed to that eevblog thread so that I would get emails on replies, but apparently not. I haven't had the time to make these requested changes yet, so I'm very glad that you took the time to contribute. I'll check the changes as soon as I can.

fredzo committed:
Moved LCD update out of interrupt rouutine to reduce delay between GPS and MCU PPS outputs.

fredzo committed:
Fixed screen refresh logic.

fredzo committed:
Moved PPS sync logic up

fredzo committed:
Update README.md

fredzo committed:
Fixed diplay update logic.
Added auto-save PWM and auto-resync feature when PPB reaches 0.
Added force sync option in PPS menu.

fredzo committed:
Merge remote-tracking branch 'origin/main'

fredzo committed:
Update README.md

fredzo committed:
Update README.md

dankar commented on src/frequency.c:
```diff
@@ -48,7 +43,7 @@ int32_t frequency_get_error()
         return 0;
     } else {
-        int32_t error = frequency - HAL_RCC_GetHCLKFreq();
+        int32_t error = frequency - 70000000 /*HAL_RCC_GetHCLKFreq()*/;
```
What's the reason for not using HAL_RCC_GetHCLKFreq? I usually prefer not to have magic numbers in the code.

fredzo replied:
Hi Daniel,
Just trying to get the code to execute faster for timing issues.
Here we avoid a function call with context switching and pushing values to the stack etc.

dankar commented on src/frequency.c:
```diff
@@ -69,3 +64,8 @@ int32_t frequency_get_ppb()
+
+bool frequency_is_stable()
+{
+    return ((num_samples == CIRCULAR_BUFFER_LEN) && (frequency_get_ppb() == 0));
```
Is 0 reasonable here? I figured it might flip between -1 and 1 as well, and we could still consider it "stable".

fredzo replied:
Well the idea is to wait long enough to have the temperature of the TCXO stabilized.
0 works well for me, the GPSDO eventually gets there when it is stable.

dankar commented on src/main.c:
```diff
@@ -67,8 +87,29 @@ void gpsdo(void)
         menu_run();
+
+        contrast_pin_state = (contrast_pwm < (contrast)) ? GPIO_PIN_RESET : GPIO_PIN_SET;
+        HAL_GPIO_WritePin(LCD_CONTRAST_GPIO_Port, LCD_CONTRAST_Pin, contrast_pin_state);
```
When I was experimenting with this myself I noticed that the contrast pin has timer alternative function, so this is possible to do with hardware PWM. Any reason you are doing it "manually"?

fredzo replied:
Yes it would be better!
I was just not experienced enough in STM32 development to do that when I started working on this and went the easier way for me!
I should probably change this to using a timer (also I don't have STM32 Cube Programmer installed so I can't update and  regenerate HAL files).

dankar commented on src/int.c:
```diff
@@ -23,30 +47,61 @@ void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef* htim)
     } else if (htim == &htim2) {
+        // TIM2 is configure for 1 second count
+        // PPS output signal
+        HAL_GPIO_WritePin(PPS_OUTPUT_GPIO_Port, PPS_OUTPUT_Pin, 1);
```
Is there no risk that we get jitter on the PPS output due to other interrupts? Is it possible to use PWM output on the pin instead?

fredzo replied:
Well I would have thought so too, but it actually works pretty well!
I don't have use for PPS output, but it was a feature request from daddygo on eevblog.
He needs a PPS output that keeps steady even if the GPS signal gets lost.
Here we sync with GPS PPS output when we get it, and we can resync if we drift too much or when we reach PPB = 0 (see parameters in PPB and PPS menu).

fredzo committed:
Update README.md

fredzo committed:
Made contrast PWM output use TIM1 instread of software.

fredzo committed:
Merge remote-tracking branch 'origin/main'

fredzo committed:
Update README.md

fredzo committed:
First step to trend screen

fredzo committed:
Next step to trend display:
- Added a new no-sat custom icon
- Added trend history storage logic
- Added vertical scale auto-range
- Fixed unset trend value management

fredzo committed:
First step to trend menu:
- Added trend display with time shift
- Added trend auto-h and auto-v scale options
- Added trend h-scale and v-scale settings

fredzo committed:
Added trend_shift value control

fredzo committed:
Fixed main trend display

fredzo committed:
Locked v-scale and h-scale edit when un auto-v or auto-h mode.

fredzo committed:
First step to h scale.

fredzo committed:
Next step to h scale

fredzo committed:
End of h-scale

fredzo committed:
Removed TODO

fredzo committed:
Added h-trend limit check on auto-h-scale.

dankar commented:
Finally got around to testing this on my device. Awesome job. I would love to merge this now. I see that you've started working on a trend screen as well, would you like to finish that before I merge?

fredzo commented:
Oh yes sorry about that!
I should have created a branch for this pull request...
Yes I'm currently working on a trend screen, and the current status of the branch is not stable.
I'll keep you posted when I'm done!

fredzo committed:
Fixed bool handling in eeprom.
Fixed menu trend order.
Fixed v-scale update.

fredzo committed:
Fixed h-scale update logic

fredzo committed:
Made h-scale precision change at 200 and 2000 rather than 100 and 1000 for better precision.

fredzo committed:
Added h-scale rounding logic.
Fixed PPM displayed value when shifting display window.

fredzo committed:
Fixed h-scale.

fredzo committed:
Added visual feedback when entering Trend sub-menu ('/' char between sat# and PPM value).
Changed mean calculation logic when h-scale > 1 to prevent values from changing when new samples arrive (use same start point for a given h-scale sample group).

fredzo committed:
Moved version to v0.1.5.
Added boot menu configuration in flash.

fredzo committed:
Update README.md

fredzo committed:
Fixed shift limit on trend navigation.
Prevent from returning to exit screen on trend menu.

fredzo committed:
Doc.

fredzo committed:
Doc

fredzo committed:
Update README.md

fredzo committed:
Update README.md

fredzo committed:
Doc

fredzo committed:
Merge remote-tracking branch 'origin/main'

fredzo commented:
Hi Daniel,
Just finished the Trend Screen, the branch is stable, you can merge it now if you want.
Best,
Frederic.

dankar commented:
Nice work! Thank you so much for contributing.
